### ISSUE TITLE
how to join my own daemon?

### ISSUE THREAD

**Issue by AlexeyRoza** [dmitigr/fcgi#4]
hello again
I need to do some tasks on the background, so how exactly i should join my own daemon in your threads?
Also i want to send my commands from console to whole thing. For ex.: delete something from DB, or find something, or execute some task, etc.

**Comment by dmitigr** [comment]
> I need to do some tasks on the background, so how exactly i should join my own daemon in your threads?

The library has no threads. You can tie the request handler with the thread pool (as shown in https://github.com/dmitigr/fcgi#hello-multithreaded-world example, where the handler is the `serve` lambda), or you can create thread for each request (so called thread-per-connection approach).

To run the process in the background you must detach it from the tty. There are tons of info around how to do it. Also, you can just use `systemd` which can do it for you automatically by using the service unit configuration like the following:
```
[Unit]
Description=example_com

[Service]
ExecStart=/path/to/fastcgi/application
Restart=on-failure
RestartSec=10s

[Install]
WantedBy=multi-user.target
```

> Also i want to send my commands from console to whole thing. For ex.: delete something from DB, or find something, or execute some task, etc.

You need to place your FastCGI application behind the HTTP server with FastCGI support, such as NGINX. Then you can use the tools like `curl` to issue HTTP requests to the HTTP server which will forward them to the FastCGI application.

**Comment by AlexeyRoza** [comment]
> The library has no threads. You can tie the request handler with the thread pool (as shown in https://github.com/dmitigr/fcgi#hello-multithreaded-world example, where the handler is the `serve` lambda), or you can create thread for each request (so called thread-per-connection approach), or you can not use threads at all -- all these depends on your needs.

my needs is permanent use of daemon program, I don't need overheat on creating new threads every conn...
The question is - how exactly do it in this situation:
    for (auto& t : threads)
      t = std::thread{serve, server.get()};

    for (auto& t : threads)
      t.join();

I see you doing double iteration - why?
Where exactly i should join my thread?

**Comment by dmitigr** [comment]
> I see you doing double iteration - why?

The first iteration creates the threads and runs the `serve` function. The `serve` function will never return since, as you can see, it is blocked by the endless loop.

The second iteration waits for each thread to complete. (In the example the threads never completes, but your application can have a condition to stop the endless loop if, for instance, some event occurred).

> Where exactly i should join my thread so it will stay there forever?

`join` must be called at the synchronization point where the application is ready to proceed after the thread(s) completes. Please, see https://en.cppreference.com/w/cpp/thread/thread/join for details

**Comment by AlexeyRoza** [comment]
shall i use this before listener starting?
std::thread foo(my_thread_routine).detach();
so it will be always working in the background?

**Comment by dmitigr** [comment]
> std::thread foo(my_thread_routine).detach();

You don't need `detach` if you're using thread pool.

> so it will be always working in the background?

It seems you're trying to make the process runs in the background. But it has nothing common with `detach`. You may want to read this to understand how to put the processes into background -- https://www.freedesktop.org/software/systemd/man/daemon.html

**Comment by AlexeyRoza** [comment]
I don't think i can use thread pool
std::vector<std::thread> threads(pool_size);
because these threads are used for http-connections...
and yes I know how to write a daemon
I just don't know how to merge my daemon with your daemon

**Comment by AlexeyRoza** [comment]
oh, there's no daemon...
so I should put all these in a daemon

**Comment by AlexeyRoza** [comment]
btw if I will run your fcgi-hellomt.cpp and then will access it through browser directly (not trough nginx port but exactly on the port which your app is listening) then i'm getting this:
terminate called after throwing an instance of 'std::runtime_error'
  what():  dmitigr::fcgi: protocol violation
Aborted (core dumped)

**Comment by dmitigr** [comment]
The example provided in `README.md` aims be to be as short as possible. You could wrap the endless loop in `serve` lambda with `try` block and catch the exceptions to prevent aborting the executable process.

**Comment by AlexeyRoza** [comment]
isn't overheat? That's try/catch for every connection...

**Comment by dmitigr** [comment]
> isn't overheat? That's try/catch for every connection...

`try` itself costs nothing. Overheads are only when throwing exceptions. But even this is not a problem.
Anyway, such a case ("protocol violation") is not realistic in production, since the FastCGI application is always behind the HTTP server and the clients has no access to the FastCGI application directly.

**Comment by AlexeyRoza** [comment]
yeah ok, just need to change 0.0.0.0 to 127.0.0.1

**Comment by dmitigr** [comment]
> yeah ok, just need to change 0.0.0.0 to 127.0.0.1

Btw, if your application on the same host as the HTTP server, you could use Unix Domain Sockets (UDS). Should be even faster.

**Comment by AlexeyRoza** [comment]
yeah almost forgot thanks
